### DISCUSSION TITLE
NetBox DNS and NetBox Branching

### DISCUSSION THREAD

**Discussion by peteeckel** [sys4/netbox-plugin-dns#389] [Announcements]
# NetBox DNS and NetBox Branching

**Do not install the Branching plugin if you are using NetBox DNS**. 

There are currently at some issues in NetBox Branching that make using it with NetBox DNS impossible. Furthermore, it is still unclear what changes have to be done within NetBox DNS to make it compatible.

 will continue to test as soon as new versions become available and remove this warning as soon as compatibility is verified. At the moment, there is no way of using both plugins together. I'm deeply sorry about this bad news because I am really looking forward to being able to use branching with NetBox DNS, but at the moment there is nothing I can do except testing and opening issues.

**Comment by wohali** [comment]
Thanks for the heads up, @peteeckel . Are links available to issues (here or against Netbox) that we can follow along to see what progress is being made? Thanks!

**Comment by peteeckel** [comment]
Unfortunately there are many open issues of which none can be picked as **the** blocking issue by itself.

When you have a look at https://github.com/netboxlabs/netbox-branching/issues you'll see what I mean. Many of the probelms stem from the fact that - like NetBox itself, and some other plugins as well - NetBox DNS relies a lot on things happening in the background when other actions are triggered.

For example, when you change a zone's nameservers, after the update to the zone itself its `NS` records need to be updated. While this seems like a trivial task, it involves a signal handler for an `m2m_changed` event to be called by Django. For this to happen in a proper database context for a branch, the handler needs to know the branch, and I currently have no idea how to get that information to it - see https://github.com/netboxlabs/netbox-branching/discussions/131 (as of yet unanswered).

The same applies to non-M2M interactions - NetBox DNS frequently updates objects when objects of other classes change, and it's necessary to pass these change operations the correct database connection for this to play well inside a branch. Unfortunately it's not really clear in which cases exactly that needs to be taken care of, and there is no proper documentation yet either. Adding this kind of code in all places where it might be necessary is nothing I would care to do, so I hope for some documentation to be provided in the near future.

Then you'll see a some issues related to custom scripts. Many of the integrations done with NetBox DNS involve custom scripts, so it's indirectly affected by the problem that scripts currently don't play well with branches.

Even internal functionality of NetBox is not safe in the context of branching - see https://github.com/netboxlabs/netbox-branching/issues/178, https://github.com/netboxlabs/netbox-branching/issues/175, https://github.com/netboxlabs/netbox-branching/issues/174, https://github.com/netboxlabs/netbox-branching/issues/155 or https://github.com/netboxlabs/netbox-branching/issues/136 for some recent examples.

Given the fact that there are still some issues with NetBox core itself when you use branching, I decided to postpone dedicated work to achieve compatibility until netbox-branching has stabilised. At the current time I regard it an early beta and not suited for production use yet.

That said, there already have been some efforts to iron out obvious no-gos with branching, most notably https://github.com/peteeckel/netbox-plugin-dns/pull/408, which removed an anti-pattern known to break with branches. More optimisations like this are planned or have already happened under the hood, so I hope NetBox DNS can be made compatible with branching without too much effort once it's stable.

It would already be a big help it it were possible to mark a plugin as branching aware, so you could have branching installed together with NetBox DNS without one breaking the other. I thought I saw an FR for this earlier but couldn't find it now, so I opened https://github.com/netboxlabs/netbox-branching/issues/182 for it.

**Comment by zeddD1abl0** [comment]
@peteeckel  - Just checking in on this discussion from 2 years ago. Is NetBox DNS still incompatible with Branching? If so, what needs to be done to bring it into operational status?

Would really like it because work is hard-sold on NetBox Branching (as much as I don't see why).

**Comment by peteeckel** [comment]
Hi @zeddD1abl0, currently NetBox DNS ist still to be considered incompatible with NetBox Branching.

However. some work has been done by NBL recently with regard to documenting requirements for plugins to support branching, which is a good thing (unfortunately I currently don't have the capacity to put in major effort, but it's on my list for this year). What's still missing is a test framework, which renders everything I will be able to commit to "best effort" - I'll implement the guidelines defined by NBL and hope it will work, basically. Without automated tests I won't declare "official" support, as I don't have a way to ensure everything works as intended. 

Next week I'll be in Munich for Autocon 5, and there I will meet some of the NBL people and discuss exactly that issue - I hope we'll find a way out of this situation together, they hinted at some improvements in the test environment, which would be great news.

**Reply by zeddD1abl0** [reply]
@peteeckel - How did your discussions go with NetBox? Is there anything I can do to help get the Branching support in? I don't quite understand how it's supposed to work, but I'm more than happy to help out.

**Reply by peteeckel** [reply]
In fact, my planned meeting with Mark didn't take place, mainly because he wasn't in MUC as originally planned.

However, and that's extremely good news, Art Hanson is preparing a PR that should solve the issues. Art is probably the person who knows the branching plugin best, and I'm confident that he is going to work it out.

The chances of a near-term solution have never been better.

**Comment by peteeckel** [comment]
Last month, @arthanson contributed code that addresses the issue, which is a major breakthrough - thank you very much, Arthur!

I released a beta release a couple of days ago, and would invite anyone interested in DNS + Branching to test it. The current version of the Beta is 1.6-beta3 and is based on the latest NetBox DNS version 1.5.11 plus Arthur's branching enhancements.  

In addition to adding the code required to support branching, Arthur also added a basic test suite, which is really awesome. I hope now we're really close to solving the issue with Branching and DNS soon.

So now is the time to start testing ... the more positive results (or fixable issues) I get, the sooner I can release 1.6.0 with support for Branching!
